Add `action_controller.allowed_redirect_hosts` configuration

Kevinrob opened the pull request:
### Summary

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

We use Rails with multiple domains (and subdomains). All these domains are configured with `config.hosts`.  
Sometimes we have dynamic redirect from one domain to an other.  
And `ActionController::Redirecting::UnsafeRedirectError` is raised.

It would be great to allow redirections to domains that the app accept in `config.hosts`.
This should not be considered as unsafe.

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

Kevinrob committed:
Allow hosts redirects from `hosts` Rails configuration

p8 commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -11,6 +11,7 @@ class UnsafeRedirectError < StandardError; end
 
     included do
       mattr_accessor :raise_on_open_redirects, default: false
+      mattr_accessor :redirect_hosts_allowed, default: []
```
As this accessor is a collection of hosts, what do you think about renaming the accessor?
```suggestion
      mattr_accessor :allowed_redirect_hosts, default: []
```

Kevinrob replied:
Yes, it's better! I changed that 👍🏻

Kevinrob commented:
@p8 Is there anything else I can do about this PR?

fiestacasey commented:
I have a similar setup and in particular one domain is in charge of authentication and redirecting to the other domain after auth. This is further complicated because the redirect is within Devise and `allow_other_host` isn't controllable without monkey patching either Rails or Devise methods. I'm sure this is the case within a lot of different gems that call `redirect_to` internally so it would be great to be able to define `allowed_redirect_hosts` in this way.

zzak commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Allow hosts redirects from `hosts` Rails configuration
```
I think providing an example in the changelog would be nice

Kevinrob replied:
What example would you like to view here?
This PR only fixes the raise of `UnsafeRedirectError` when redirecting to an allowed hosts from `config.hosts`.

zzak replied:
```ruby
redirect_to("http://other_domain.com/") #=> UnsafeRedirectError

Rails.application.config.action_controller.allowed_redirect_hosts << "other_domain.com"
redirect_to("http://other_domain.com/") #=> 302
```

zzak commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -198,6 +199,7 @@ def _url_host_allowed?(url)
         host = URI(url.to_s).host
 
         return true if host == request.host
+        return true if ActionDispatch::HostAuthorization::Permissions.new(allowed_redirect_hosts).allows?(host)
```
This looks like an anti-pattern to me, like the list of allowed hosts should be known before we get here or evaluated lazily. Might just be me though 🤔

zzak replied:
Is it possible to initialize this sooner?

Kevinrob replied:
Like that https://github.com/rails/rails/pull/45680/commits/0e5171de5e592468d8c688b97ec2fbdc277234f4 ?

tom-lord commented:
Any update on this feature?

We've run into the same issue, where we frequently want to redirect between two subdomains. Currently the only option is to disable this security feature and verify manually that the URL is safe.

Additionally, what's odd is that [the `url_from` helper automatically considers other subdomains "safe"](https://api.rubyonrails.org/v7.0.7/classes/ActionController/Redirecting.html#method-i-url_from) meanwhile `UnsafeRedirectError` will still be raised when redirecting to another subdomain.

owst commented:
Hi @Kevinrob, thanks for getting started with this PR, would you be able to get it across the line? There are a couple of sensible comments from @zzak and a build failure - hopefully they are both straightforward to address?

We'd like to see this added to Rails as we are also hitting an unsafe redirect during our authentication flow in the same way as https://github.com/apokalipto/devise_saml_authenticatable/issues/237 - adding our identity provider's hostname to `allowed_redirect_hosts` would be a preferred fix

Kevinrob commented:
> Hi @Kevinrob, thanks for getting started with this PR, would you be able to get it across the line? There are a couple of sensible comments from @zzak and a build failure - hopefully they are both straightforward to address?
> 
> We'd like to see this added to Rails as we are also hitting an unsafe redirect during our authentication flow in the same way as [apokalipto/devise_saml_authenticatable#237](https://github.com/apokalipto/devise_saml_authenticatable/issues/237) - adding our identity provider's hostname to `allowed_redirect_hosts` would be a preferred fix. There is some discussion (but no conclusion) on a Devise issue ([heartcombo/devise#5462](https://github.com/heartcombo/devise/pull/5462)) and it seems quite a few people are similarly affected, with a handful of alternative approaches/monkey patches being proposed - it would be great if there was an official Rails fix.

Hi @owst, I just rebased the PR against `main` and the tests is 🟢 (except one strange fail that I thing is not related).
I don't know how to bring more visibility to this PR.

Kevinrob commented:
> > Hi @Kevinrob, thanks for getting started with this PR, would you be able to get it across the line? There are a couple of sensible comments from @zzak and a build failure - hopefully they are both straightforward to address?
> > We'd like to see this added to Rails as we are also hitting an unsafe redirect during our authentication flow in the same way as [apokalipto/devise_saml_authenticatable#237](https://github.com/apokalipto/devise_saml_authenticatable/issues/237) - adding our identity provider's hostname to `allowed_redirect_hosts` would be a preferred fix. There is some discussion (but no conclusion) on a Devise issue ([heartcombo/devise#5462](https://github.com/heartcombo/devise/pull/5462)) and it seems quite a few people are similarly affected, with a handful of alternative approaches/monkey patches being proposed - it would be great if there was an official Rails fix.
> 
> Hi @owst, I just rebased the PR against `main` and the tests is 🟢 (except one strange fail that I thing is not related). I don't know how to bring more visibility to this PR.

@rafaelfranca Maybe you can help us on this?

p8 commented:
@Kevinrob Could you try rebasing to main again? Seems like there were some CI issues.

p8 commented on actionpack/test/controller/redirect_test.rb:
```diff
@@ -617,6 +617,16 @@ def test_redirect_to_instrumentation
+      end
+    end
+  end
+
```
This also needs a test where the url isn't in the allowed redirect hosts.

Kevinrob replied:
Thanks, I added that 👍

Kevinrob commented:
> @Kevinrob Could you try rebasing to main again? Seems like there were some CI issues.

Yes, it's good now 👌

nicowenterodt commented:
Just stumbled upon this one. For a multi-domain environment this would be really helpful to have. Thanks @Kevinrob bringing this one up 👍

Kevinrob commented:
@p8 I added the test as you ask in your last review ✅ Do we need to do something else on this PR?

sbiastoch commented:
Hi all, I also just came across this over my Rails 7 migration. Very happy to see there is already a PR in the pipeline addressing this issue!

I have a small suggestion to further improve this setup: If the originally requested URL would be available from `ActionController::Redirecting::UnsafeRedirectError`, one could leverage `rescue_from` to apply some custom logic to either re-throw the exception or dismiss it. This addition would nicely complement this PR in my opinion :-)

Kevinrob commented:
This PR is waiting for reviews since more than 2 years 😞
A lot of peoples added a 👍 and comments for having this.

Anyone can help us on this PR?
@eileencodes @tenderlove maybe?

Kevinrob commented:
@p8 Sorry to insist, what more can I do about this PR?

byroot commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -248,5 +250,9 @@ def _ensure_url_is_http_header_safe(url)
       end
+
+      def _allowed_redirect_hosts_permissions
+        @allowed_redirect_hosts_permissions ||= ActionDispatch::HostAuthorization::Permissions.new(allowed_redirect_hosts)
```
`ActionDispatch::HostAuthorization::Permissions.new` compile regexps, hence we shouldn't create one per request.

byroot commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -232,6 +233,7 @@ def _url_host_allowed?(url)
         host = URI(url.to_s).host
 
         return true if host == request.host
+        return true if _allowed_redirect_hosts_permissions.allows?(host)
         return false unless host.nil?
         return false unless url.to_s.start_with?("/")
```
We should test the happy path like `host.nil` first.

byroot reviewed (commented):
The feature makes sense to me, but I'm concerned about recompiling, potentially many regexp, on every request that does a redirect.

Compiling regexps is expensive. So you should build it only when the config change

Also the `redicted_to` "Open Redirect protection" documentation section need to be updated.

byroot committed:
Avoid rebuilding HostAuthorization::Permissions over and over

Compiling regexp is a very CPU intensive process, you don't
want to do that for the same value again and again.

byroot commented:
I wanted to cleanup your PR myself, but it seems you disabled repo owner push into PRs, so here it is: https://github.com/rails/rails/compare/main...byroot:rails:allow_redirect_from_configuration_hosts

If you update the documentation that will be good for me.

byroot commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -248,5 +250,9 @@ def _ensure_url_is_http_header_safe(url)
       end
+
+      def _allowed_redirect_hosts_permissions
+        @allowed_redirect_hosts_permissions ||= ActionDispatch::HostAuthorization::Permissions.new(allowed_redirect_hosts)
```
I could see `config.hosts` containing some IPs in case you have some server without a name. It's niche but I don't really see a reason to prevent it?

p8 replied:
Yeah, can't think of a possible issue, but maybe `Permissions` should be moved out of `ActionDispatch::HostAuthorization` as it is used more generically now?

p8 commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -31,6 +32,10 @@ class Railtie < Rails::Railtie # :nodoc:
     end
 
+    initializer "action_controller.set_allowed_redirect_hosts" do |app|
+      config.action_controller.allowed_redirect_hosts << app.config.hosts
```
This is a breaking change.
An application might have defined an old unused hostname in `config.hosts`.
With this change the redirects are allowed to that old hostname, without the developer knowing.

Kevinrob replied:
You're right! I changed the code for not initializing this config with `app.config.hosts` 👍🏻

p8 commented:
`allowed_redirect_hosts` should probably be documented in the Configurations guide as well:
https://edgeguides.rubyonrails.org/configuring.html#configuring-action-controller

Kevinrob committed:
Don't initialize allowed_redirect_hosts with hosts

Kevinrob committed:
Add documentation about allowed_redirect_hosts

Kevinrob commented:
> I wanted to cleanup your PR myself, but it seems you disabled repo owner push into PRs, so here it is: [main...byroot:rails:allow_redirect_from_configuration_hosts](https://github.com/rails/rails/compare/main...byroot:rails:allow_redirect_from_configuration_hosts)
> 
> If you update the documentation that will be good for me.

Thank you @byroot !
I don't have the option to allow repo owner to push into this PR 🤔 But I cherry-pick the commit, thanks!

The documentation was added too 👍 

By the way, I changed for not initializing `config.action_controller.allowed_redirect_hosts` with `hosts` for not having a breaking changes.

zzak commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -32,6 +33,10 @@ class Railtie < Rails::Railtie # :nodoc:
     end
 
+    initializer "action_controller.set_allowed_redirect_hosts" do |app|
+      config.action_controller.allowed_redirect_hosts = []
```
I think you want to set the class attribute in the initializer, you've already made the config an empty array above (line 19). Right?

zzak commented on guides/source/configuring.md:
```diff
@@ -2021,6 +2021,10 @@ The default value depends on the `config.load_defaults` target version:
 
+#### `config.action_controller.allowed_redirect_hosts`
+
+Specifies a list of allowed hosts for redirects. `redirect_to` will allow redirects to them without raising an error.
```
```suggestion
Specifies a list of allowed hosts for redirects. `redirect_to` will allow redirects to them without raising an `UnsafeRedirectError` from error.
```

zzak commented:
Can you also update the documentation for `redirect_to` and friends, and [squash your commits](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits) when you're ready? :bow:

p8 commented on actionpack/CHANGELOG.md:
```diff
@@ -223,4 +223,8 @@
 
     *Petrik de Heus*
 
+*   Allow hosts redirects from `hosts` Rails configuration
```
Changelog entries need to go to the top.
